wren-ai-service/src/pipelines/generation/chart_generation.py

```diff
@@ -0,0 +1,278 @@
+            "schema": ChartGenerationResults.model_json_schema(),
+        },
+    }
+}
+
+
+class ChartGeneration(BasicPipeline):
+    def __init__(
+        self,
+        llm_provider: LLMProvider,
+        **kwargs,
+    ):
+        self._components = {
+            "prompt_builder": PromptBuilder(
+                template=chart_generation_user_prompt_template
+            ),
+            "generator": llm_provider.get_generator(
+                system_prompt=chart_generation_system_prompt,
+                generation_kwargs=CHART_GENERATION_MODEL_KWARGS,
+            ),
+            "chart_data_preprocessor": ChartDataPreprocessor(),
+            "post_processor": ChartGenerationPostProcessor(),
+        }
+        self._configs = {
+            "vega_schema": requests.get(
+                "https://vega.github.io/schema/vega-lite/v5.json"
+            ).json(),
+        }
+
+        super().__init__(
+            AsyncDriver({}, sys.modules[__name__], result_builder=base.DictResult())
+        )
+
+    def visualize(
+        self,
+        query: str,
+        sql: str,
+        data: dict,
+        language: str,
+    ) -> None:
+        destination = "outputs/pipelines/generation"
+        if not Path(destination).exists():
+            Path(destination).mkdir(parents=True, exist_ok=True)
+
+        self._pipe.visualize_execution(
+            ["post_process"],
+            output_file_path=f"{destination}/chart_generation.dot",
+            inputs={
+                "query": query,
+                "sql": sql,
+                "data": data,
+                "language": language,
+                **self._components,
+                **self._configs,
+            },
+            show_legend=True,
+            orient="LR",
+        )
+
+    @async_timer
+    @observe(name="Chart Generation")
+    async def run(
+        self,
+        query: str,
+        sql: str,
+        data: dict,
+        language: str,
+    ) -> dict:
+        logger.info("Chart Generation pipeline is running...")
+        return await self._pipe.execute(
+            ["post_process"],
+            inputs={
+                "query": query,
+                "sql": sql,
+                "data": data,
+                "language": language,
+                **self._components,
+                **self._configs,
+            },
+        )
+
+
+if __name__ == "__main__":
+    from langfuse.decorators import langfuse_context
+
+    from src.core.engine import EngineConfig
+    from src.core.pipeline import async_validate
+    from src.providers import init_providers
+    from src.utils import init_langfuse, load_env_vars
+
+    load_env_vars()
+    init_langfuse()
+
+    llm_provider, _, _, engine = init_providers(EngineConfig())
+    pipeline = ChartGeneration(
+        llm_provider=llm_provider,
+    )
+
+    pipeline.visualize(query="", sql="", data={})
+    async_validate(lambda: pipeline.run(query="", sql="", data={}))
+
+    langfuse_context.flush()
```

Comment:
how about use dry_run_pipeline function to instead of here?

Reply:
fixed